metagraph/src/annotation/taxonomy/tax_classifier.hpp

```diff
@@ -0,0 +1,191 @@
+     * @param [input] label_taxid_map_filepath ->  path to acccession version to taxid lookup table (".accession2taxid" file). Mandatory if the label doesn't contain the 'taxid'.
+     */
+    TaxonomyClsAnno(const graph::AnnotatedDBG &anno,
+                    const double lca_coverage_rate,
+                    const double kmers_discovery_rate,
+                    const std::string &tax_tree_filepath,
+                    const std::string &label_taxid_map_filepath = "");
+    TaxonomyClsAnno() {};
+    ~TaxonomyClsAnno() {};
+
+    // todo implement
+    void export_taxdb(const std::string &filepath) const;
+
+    // todo implement
+    TaxId assign_class(const std::string &sequence) const;
+
+PRIVATE_TESTABLE:
+    /**
+     * Reads and returns the taxonomic tree as a list of children.
+     *
+     * @param [input] tax_tree_filepath -> path to a "nodes.dmp" file.
+     * @param [output] tree -> tree stored as a list of children.
+     */
+    void read_tree(const std::string &tax_tree_filepath,
+                   ChildrenList *tree);
+
+    /**
+     * rmq_preprocessing computes 'this->rmq_data', 'this->precalc_log' and 'this->precalc_pow2' fields.
+     *
+     * @param [input] tree_linearization -> the linearization of the taxonomic tree.
+     */
+    void rmq_preprocessing(const std::vector<TaxId> &tree_linearization);
+
+    /**
+     * dfs_statistics method calculates the following fields:
+     *      + tree_linearization;
+     *      + this->node_depth;
+     *      + this->node_to_linearization_idx.
+     *
+     * @param [input] node -> the node that is currently processed.
+     * @param [input] tree -> the taxonomic tree stored as a list of children.
+     * @param [output] tree_linearization -> the linearization of the received tree.
+     */
+    void dfs_statistics(const TaxId node,
+                        const ChildrenList &tree,
+                        std::vector<TaxId> *tree_linearization);
+
+    /**
+     * rmq_data[0] contains the taxonomic tree linearization
+     *          (e.g. for root 1 and edges={1-2; 1-3}, the linearization is "1 2 1 3 1").
+     * rmq_data[l][x] returns the node with the maximal depth among positions [x, x+2^l-1] in the linearization
+     *          (e.g. rmq_data[3][6] return the node with max depth in [6, 13]).
+     */
+    std::vector<std::vector<TaxId>> rmq_data;
+
+    /**
+     * node_to_linearization_idx[node] returns the index of the first occurrence of node
+     * in the tree linearization order. This array will be further used inside a RMQ query.
+     */
+    tsl::hopscotch_map<TaxId, uint32_t> node_to_linearization_idx;
+
+    /**
+     * fast_log2 is a table for a fast compute of log2(x).
+     */
+    std::vector<uint32_t> fast_log2;
+
+    /**
+     * fast_pow2 is a table for a fast compute of pow2(x).
+     */
+    std::vector<uint32_t> fast_pow2;
+
+    const graph::AnnotatedDBG *_anno_matrix = NULL;
```

Comment:
do you need this class to be copy assignable (`operator=`)? if not, then it's cleaner to make this a const reference instead

Reply:
This  class has an empty constructor `TaxonomyClsAnno() {};` that is used for all the unit tests. If i make the `_anno_matrix` a const reference, then I get a compilation error because I am not assigning it inside this empty class constructor. Is there any other better way to implement it?

Reply:
private members have underscore in the end and not in the beginning
```suggestion
    const graph::AnnotatedDBG *anno_matrix_ = NULL;
```

Reply:
Why do you need this as a member, anyway? It's used only in that constructor.